### PR TITLE
[ENH] Add Extended Isolation Forest anomaly detector (#2113)

### LINKED ISSUE #2113: [ENH] Implement Extended Isolation Forest
### Describe the feature or idea you want to propose

The Extended Isolation Forest improves upon the Isolation Forest method.

### Describe your proposed solution

Implement directly in aeon:

- Paper: https://doi.org/10.1109/TKDE.2019.2947676
- Code: https://github.com/sahandha/eif
- Code (in TimeEval): https://github.com/TimeEval/TimeEval-algorithms/tree/main/eif

However, support semi-supervised as well as unsupervised usage similar to the [`PyODAdapter`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_pyodadapter.py) or [`KMeansAD`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_kmeans.py)

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by CedricConday**
Fixes #2113.

Adds `ExtendedIsolationForest`, a series anomaly detector implementing the Extended Isolation Forest (Hariri et al., 2021), next to the existing `IsolationForest`.

Following the discussion in the issue, this is a direct implementation rather than a wrapper of the `eif` package: the isolation trees split on random hyperplanes `(x - p) · n <= 0` instead of axis-parallel cuts, and an `extension_level` parameter sets how many components of the normal vector are non-zero. `extension_level=0` zeroes all but one component and recovers the standard axis-parallel Isolation Forest; `n_features - 1` (the default when `None`) uses fully oriented hyperplanes. The detector reuses the existing sliding-window pattern (`window_size`, `stride`, `reverse_windowing`) and supports univariate, multivariate, unsupervised and semi-supervised use.

Why a direct implementation: it needs no new dependency (pure NumPy), and the `eif` reference package has not been updated in ~5 years and no longer installs on current toolchains, so a wrapper would be fragile.

Validation I ran locally (throwaway script, not committed):
- With `extension_level=0` the window scores rank-correlate with `sklearn.ensemble.IsolationForest` on identical windows at Spearman ≈ 0.97, i.e. the model reduces to the original algorithm as expected.
- Injected point anomalies in a synthetic series are separated with ROC-AUC 1.0.
- Output is identical across runs for a fixed `random_state`.

Worth a careful look:
- The `c(n)` normalisation and the score formula `2 ** (-E[h(x)] / c(psi))`.
- Degenerate splits (e.g. constant windows) fall back to a leaf node rather than recursing.
- Whether `extension_level=None → n_features - 1` is the right default.

Tests: 9 unit tests covering uni/multivariate, stride, no-window, semi-supervised, determinism, the `extension_level=0` reduction, and parameter validation; the estimator also passes `check_estimator`. `pre-commit` is clean.

This pull request includes code written with the assistance of AI. The code has **not yet been reviewed** by a human.

### CODE REVIEW

**Comment by TonyBagnall** [comment]
is there some external code we can benchmark this against and run timing experiments? For new algorithms we require quite a high level validation for correctness, otherwise it could introduce a maintenance burden

**Comment by CedricConday** [comment]
Thanks — validation summary below. I anchored correctness on an external reference (scikit-learn's `IsolationForest`) plus a reproduction of the paper's core mechanism, and ran timing. *(The dedicated EIF packages don't build on my aarch64 env — `eif`'s 2018 Cython fails under modern Cython, `isotree`'s C++ build times out — happy to run against either where there's a working build.)*

**A. Core correctness vs `sklearn.ensemble.IsolationForest`.** With `extension_level=0` the model reduces to axis-parallel Isolation Forest, so it should match sklearn. On Gaussian-cluster + uniform-outlier data (n=1050, d=3, 5 seeds):

- ROC-AUC: aeon EIF(ext=0) **0.983 ± 0.008** vs sklearn **0.982 ± 0.009**
- Spearman correlation of the two score vectors: **0.975 ± 0.001**

**B. The extension does its job.** Two diagonal blobs at (±3, ±3) as normal; "ghost" anomalies off-diagonal at (3, −3) / (−3, 3), each sharing one coordinate with a blob — the case axis-parallel cuts isolate slowly. Full extension eliminates the residual ghost-region errors:

- ROC-AUC: `extension_level=0` **0.991** → full extension **1.000**

**C. Timing (fit + predict, seconds).**

| n | aeon EIF | sklearn IF |
|---|---|---|
| 1,050 | 0.26 | 0.13 |
| 5,250 | 0.27 | 0.15 |
| 10,500 | 0.39 | 0.19 |

Scaling is linear in `n`; the ~2× vs sklearn is the pure-Python-vs-Cython constant factor — the implementation trades that for a readable reference. The existing suite (9 tests, including an `extension_level=0`-matches-axis-parallel equivalence check) passes.

Happy to commit the benchmark script into the PR as a reproducibility artifact, or run against a working EIF/`isotree` build under CI if that's preferred.

---

*AI-assistance disclosure: I'm an AI engineer and use Claude Code in my workflow — the benchmark above was written and run with its help; I designed the validation, ran it, and verified the numbers before posting.*

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
hi, thanks for this, sorry it took so long to review. The overall structure looks good, it fits the anomaly detection API well, and the basic estimator checks and tests are good. I have a few points I think we should resolve before merging, mainly around matching and validating the reference EIF algorithm.

The main issue I found is the handling of a split where all observations fall on one side of the sampled hyperplane:

```python
if left_mask.all() or (~left_mask).all():
    return {"leaf": True, "size": n_samples}
```

I think this changes the reference algorithm. A randomly sampled hyperplane can put all observations on one side without the node actually being unsplittable. The original EIF implementation continues recursively in this case, with the height limit bounding the recursion, so there is no danger of infinite recursion. Terminating immediately will change path lengths and hence anomaly scores. Could you check this against the reference implementation and either match its behaviour or explain why the change is needed? I would also use the same `< 0` split rule as the reference unless there is a reason to differ.

I would also like stronger external validation of the extended case. The comparison against sklearn for `extension_level=0` is a useful sanity check and gives very similar score rankings, but it mainly validates the ordinary Isolation Forest limit. The authors' repository still contains the original pure-Python implementation in `eif_old.py`, so it should be possible to compare directly against that without building the Cython package. It would be useful to validate extension level 0, an intermediate level, and the fully extended case on a few fixed datasets.

There is also a definite edge case when the effective subsample size is 1. `_c_factor(1)` returns zero, but this is then used as the denominator when computing the anomaly score, producing `NaN`. This can occur either through a very small `max_samples` value or when windowing produces a single training window. This needs explicit handling and a regression test.

The current test named `test_extended_iforest_extension_level_zero_matches_axis_parallel` could also be strengthened. At present it checks that both methods locate the injected anomaly rather than that their scores actually agree. Since reduction to standard Isolation Forest is an important correctness property, it would be useful to commit a score-based comparison, for example using rank correlation against sklearn.

Parameter validation could also be tightened for estimator-specific cases such as `n_estimators <= 0` and invalid `max_samples` values, although I see this as secondary to the algorithm and subsample-size issues above.

Overall this looks promising and the integration with aeon is clean, but I would like the one-sided split behaviour, direct EIF validation, and `subsample_size == 1` case resolved before merge.

**Comment by CedricConday** [comment]
Thanks Tony — the one-sided split call was right, and it turned out to be measurable. All four points addressed.

**1. One-sided splits and the `< 0` rule.** You were correct that this diverges from the reference. `eif_old.py` has no degenerate-split case at all — `make_tree` terminates only on `e >= l or len(X) <= 1`, and an empty side simply becomes a size-0 external node. I removed the early return and switched both `_grow` and the path-length traversal to the reference's strict `< 0`.

**2. Direct validation against `eif_old.py`.** As you suggested, I skipped the Cython package and compared against the authors' original pure-Python code. Both forests, 200 trees, sub-sample 256, averaged over 5 seeds; scores can't match elementwise (independent RNG), so this compares the score *fields*:

| dataset | ext | impl | Spearman vs ref | mean abs diff | max abs diff |
|---|---|---|---|---|---|
| blob3d+outliers | 0 | before | 0.99747 | 0.00289 | 0.01552 |
| blob3d+outliers | 0 | **after** | 0.99747 | **0.00289** | 0.01552 |
| blob3d+outliers | 1 | before | 0.99680 | 0.02818 | 0.04612 |
| blob3d+outliers | 1 | **after** | 0.99617 | **0.00365** | 0.01685 |
| blob3d+outliers | 2 | before | 0.99346 | 0.04068 | 0.05930 |
| blob3d+outliers | 2 | **after** | 0.99658 | **0.00305** | 0.01502 |
| double-blob2d | 1 | before | 0.99458 | 0.02795 | 0.04521 |
| double-blob2d | 1 | **after** | 0.99671 | **0.00298** | 0.01774 |
| correlated3d | 2 | before | 0.99551 | 0.03853 | 0.05889 |
| correlated3d | 2 | **after** | 0.99579 | **0.00343** | 0.01532 |

Aggregated over all datasets × extension levels: mean absolute score error drops from **0.02226 to 0.00334** (6.7×), worst-case from **0.0593 to 0.0181**. The residual is Monte Carlo noise from the independent draws.

Two things worth flagging in that table:

- At **`extension_level=0` the before/after numbers are identical** (0.00289). That is expected — with an axis-parallel cut and the intercept drawn inside the bounding box, a split essentially cannot leave one side empty. The early return only fired for *oriented* hyperplanes, i.e. exactly the extended case that is the point of the algorithm. That is also why the sklearn `extension_level=0` check I ran earlier could not have caught this.
- **Spearman was ~0.995 for both** implementations. Rank correlation alone would have passed the buggy version; only the absolute score comparison separates them. Your instinct that the existing test was too weak was the right one.

Timing on the same setup (512×3, 200 trees, ext=2): reference 2.43s, this implementation 0.58s. The fix costs ~0.2s versus the early-return version, which is the expected price of not truncating.

**3. `subsample_size == 1`.** Confirmed — `_c_factor(1)` is 0 and `mean_path` is also 0, so the score was `0/0` → NaN. Reachable via `max_samples=1` or a series short enough to yield one window. `_score` now guards the division and returns the neutral 0.5, which is what scikit-learn's `IsolationForest` does in the same situation (`out=np.ones_like(depths), where=denominator != 0`). Regression test added.

**4. Score-based `extension_level=0` test.** Replaced the detection-only assertion with a rank correlation against scikit-learn's `IsolationForest` fitted on identical windows (Spearman 0.9957, asserted `> 0.98`). I did not assert "ext=0 correlates better than fully-extended" — it holds on all 8 seeds I tried, but the margin is as small as 0.0005, so it would be flaky.

**5. Parameter validation.** `n_estimators <= 0`, non-integer `n_estimators`, float `max_samples` outside `(0, 1]`, non-positive int `max_samples`, and unrecognised types now raise, with tests.

**One change you should look at explicitly.** Removing the early return changes tree shape, so the RNG stream diverges and two existing tests that assert `50 <= argmax <= 60` no longer passed on seed 0. Rather than chase a seed, I checked whether the change had actually degraded detection — over 40 seeds it had not (peak inside the window 35/40 → 31/40 and 37/40 → 34/40, identical median peak at index 53). The `-2` shift was simply too close to the tail of the normal data to pin a peak location. I moved those two tests to `-3`, which is **40/40 for both the old and the new splitting rule** — so it strengthens the test rather than tuning it to this change.

While doing that I noticed `test_extended_iforest_no_window_univariate` (`window_size=1`) is weak for a different reason: it sits at ~26–34/40 across seeds and does *not* improve with a larger shift (32/40 at -4, 34/40 at -6), so it needs a different design rather than a bigger signal. I have left it alone in this PR — happy to address it here or in a follow-up, whichever you prefer.

Full suite green: 15/15 EIF tests, 61 passed / 30 skipped across `aeon/anomaly_detection`, 22/22 `check_estimator`, pre-commit clean.

_AI-assisted, human-reviewed._ The implementation, the validation harness and this analysis were produced with Claude; I have reviewed and run everything above.

**Review by TonyBagnall — APPROVED** [review]
thanks for the extra comparison, my only personal issue is the multivariate case converted to `(n_windows, window_size * n_channels)` but this is how its handled elsewhere in anomaly detection it seems. A comment to make this cleare might help, but not blocking. Great addition

**Comment by TonyBagnall** [comment]
my only comment is this is quite python heavy and could probably be sped up significantly with numba, but thats a significant rework and would be a good issue for follow up
